_posts/2021/2021-08-02-dont-call-replacement-classes-new.md

```diff
@@ -0,0 +1,95 @@
+---
+title: "Don't call a new version of something 'new' when refactoring"
+description: "A opinionated naming subtlety"
+layout: article
+category: ruby
+image:
+  base: '2021/dont-call-replacement-classes-new'
+  alt: "New"
+  credit: Nick Fewings
+  source: "https://unsplash.com/photos/1SsUquHPNT8"
+
+---
+
+In long-lived applications, we often want to rewrite or refactor an approach that we have previously taken.
+
+Here's a, non-obvious, naming technique I've used during refactoring to ensure that the improved code ends up in well named classes and any old implementation is clearly marked for deletion.
+
+You might find this helps organise your work or communicate the intent of future changes with your teammates.
+
+
+## Instead of...
+
+...littering your codebase with the word "new":
+
+```ruby
+class NewDoThingJob << ActiveJob
+  # new implementation
+end
+```
+
+or
+
+```ruby
+class DoThingJobVersionTwo << ActiveJob
+  # new implementation
+end
+```
+
+
+## Choose...
+
+to rename or, better, namespace the existing code to mark it as deprecated. Then create the replacement version with the correct future naming.
+
+```ruby
+module ToBeRemoved
+  class DoThingJob << ActiveJob
+    # existing implementation
+  end
+end
+```
+
+```ruby
+class DoThingJob < ToBeRemoved::DoThingJob
+```
+
+Then change all the places the non-namespaced version is called. Then deploy and ensure that no jobs are still enqueued with the non-namespaced version.
+
+Now you're free to reimplement in the correct (and future) location.
+
+```ruby
+class DoThingJob
+  # new implmentation
+end
+```
+
+Once you've built, tested and changed all the calls to the new version you can delete the (now unused) namespaced version.
+
+
+## Why?
+
+This is a habit that helps communicate intent.
+
+These refactoring projects are often large in scale and it can be hard to see what's going on in the midst of a series of changes. But this methodology clearly lays out the intention to replace this existing code.
```

Comment:
Remove the 'but'.

*intention to replace existing code

Reply:
```suggestion
These refactoring projects are often large in scale and it can be hard to see what's going on in the midst of a series of changes. This methodology clearly lays out the intention to replace existing code.
```